### PR TITLE
Add support for privacy API

### PR DESCRIPTION
**Pull request by shadowspawn**
Add the privacy API, for GDPR compliance et al: https://docs.moodle.org/dev/Privacy_API

Update required version of Moodle to oldest LTS version.

### CODE REVIEW

File: classes/privacy/provider.php [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        // Describe the data exported to kuraCloud.
+        $collection->add_external_location_link('kuracloud_sync', [
+            'firstname' => 'privacy:metadata:kuracloud_sync:firstname',
+            'lastname' => 'privacy:metadata:kuracloud_sync:lastname',
+            'idnumber' => 'privacy:metadata:kuracloud_sync:idnumber',
+            'email' => 'privacy:metadata:kuracloud_sync:email',
+        ], 'privacy:metadata:kuracloud_sync');
+
+        return $collection;
+    }
```
Comment (victoriabroad-adi) [inline classes/privacy/provider.php:103]: Why doesn't this one get a function comment?

File: classes/privacy/provider.php [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+function kuracloud_delete_stale_data_for_user(int $userid) {
+    global $DB;
+
+    // Get list of stale records, which are no longer associated with a course.
+    $sql = "SELECT kc_users.id
+              FROM {block_kuracloud_users} kc_users
+         LEFT JOIN {block_kuracloud_courses} kc_courses on kc_users.remote_instanceid = kc_courses.remote_instanceid AND kc_users.remote_courseid = kc_courses.remote_courseid
+         LEFT JOIN {course} c on c.id = kc_courses.courseid
+             WHERE c.id IS NULL AND kc_users.userid = :userid
+    ";
```
Comment (victoriabroad-adi) [inline classes/privacy/provider.php:55]: What happens when a user is part of a dead course and part of an alive course?

File: classes/privacy/provider.php (hunk 1 above)
Comment (shadowspawn) [reply classes/privacy/provider.php:103]: Huh. Will do.

File: classes/privacy/provider.php (hunk 2 above)
Comment (shadowspawn) [reply classes/privacy/provider.php:55]: There is a row per user per course, so this only finds the rows where the course is gone burgers.

File: lang/en/block_kuracloud.php [hunk 3]
```diff
@@ -69,4 +69,14 @@
...
+$string['remotecoursedeleted'] = 'Error: The mapped kuraCloud course has been deleted - Please restore the course or map this course to a different kuraCloud course.';
+
+$string['privacy:metadata:block_kuracloud_users'] = 'Information to match Moodle users with kuraCloud users';
+$string['privacy:metadata:block_kuracloud_users:userid'] = 'Moodle user id';
+$string['privacy:metadata:block_kuracloud_users:remote_studentid'] = 'kuraCloud student id';
+
+$string['privacy:metadata:kuracloud_sync'] = 'In order to sync kuraCloud with Moodle, user data needs to be sent to kuraCloud';
+$string['privacy:metadata:kuracloud_sync:firstname'] = 'Your first name is sent to kuraCloud to allow a better user experience';
+$string['privacy:metadata:kuracloud_sync:lastname'] = 'Your last name is sent to kuraCloud to allow a better user experience';
+$string['privacy:metadata:kuracloud_sync:idnumber'] = 'Your Moodle idnumber is sent to help match up with your Moodle account, and help manage student accounts in kuraCloud';
```
Comment (victoriabroad-adi) [inline lang/en/block_kuracloud.php]: `idnumber` or id number ?

File: classes/privacy/provider.php [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+                 WHERE kc_users.userid = :userid
+        ";
+        $params = [
+            'userid' => $userid,
+        ];
+        $details = $DB->get_records_sql($sql, $params);
+
+        foreach ($contextlist->get_contexts() as $context) {
+            $courseid = $context->instanceid; // Tentative, as also need to check contextlevel.
+            if ($context->contextlevel === CONTEXT_COURSE && array_key_exists($courseid, $details)) {
```
Comment (victoriabroad-adi) [inline classes/privacy/provider.php:243]: Can't straight up delete this as this function call is needed? `array_key_exists`

File: lang/en/block_kuracloud.php (hunk 3 above)
Comment (shadowspawn) [reply lang/en/block_kuracloud.php]: I'll change to "id number" to match the others, I didn't say "firstname".

File: classes/privacy/provider.php (hunk 4 above)
Comment (shadowspawn) [reply classes/privacy/provider.php:243]: Yes, and was easier and more in style of other routines to loop than try and build into sql query.

**Comment by victoriabroad-adi** [comment]
Looks good to me 👍 

Very readable

File: classes/privacy/provider.php (hunk 1 above)
Comment (shadowspawn) [reply classes/privacy/provider.php:103]: Added

File: lang/en/block_kuracloud.php (hunk 3 above)
Comment (shadowspawn) [reply lang/en/block_kuracloud.php]: Changed

**Pull request merged**